### ISSUE TITLE
Multiple `takeLatest` in saga root ?

### ISSUE THREAD

**Issue by Grmiade** [yelouafi/redux-saga#124]
I can't have several `takeLatest` in my saga root, it's normal ?

```javascript
export default function* root() {
  yield* takeLatest(Actions.LOGIN, login);
  yield* takeLatest(Actions.LOGOUT, logout);
}
```

Only the one `takeLatest` is take into account :/

**Comment by zalmoxisus** [comment]
@Grmiade, you should first fork them, otherwise the second `yield` will not be reached:
```js
export function* watchLogin() {
  yield* takeLatest(Actions.LOGIN, login);
  yield* takeLatest(Actions. LOGOUT, logout);
}

export function* watchLogout() {
  yield* takeLatest(Actions. LOGOUT, logout);
}

export default function* root() {
  yield [
    fork(watchLogin),
    fork(watchLogout)
  ];
}
```

The same goes for `takeEvery`.

**Comment by yelouafi** [comment]
@Grmiade 
`yield*` allows only for sequential compositions; see http://yelouafi.github.io/redux-saga/docs/advanced/SequencingSagas.html

If you want to yield muliple tasks in parallel (note `yield` and not `yield*`)
 
```javascript
export default function* root() {
  yield [
     takeLatest(Actions.LOGIN, login),
     takeLatest(Actions.LOGOUT, logout)
  ]
}
```

I works because `takeLatest` and `takeEvery` return Iterators, and `redux-saga` can also handle yielding iterators directly. But this has a minor side effect on monitoring Sagas (the Saga monitor will print `anonymous` on the places you call `yield`); 

If you've worked before with languages like C/C++ you can view the difference between direct call (`yield iterator`) and delegation (`yield* iterator`) like the difference between calling a function and a *marco* in your code.

Using `yield takeLatest` is like calling a function, the middleware will run the Iterator and resumes the parent Saga when the iterator terminates (in the above case the parent Saga will resume when all the child iterators in the array terminates)

Using delegation (`yield*`) the Iterator will be *spread* into the parent Saga: It's like if you replaced the body of the parent Saga by the instructions of the invoked iterator. So in you code example

```javascript
export default function* root() {
  yield* takeLatest(Actions.LOGIN, login);
  yield* takeLatest(Actions.LOGOUT, logout);
}
```

The body of `root` will run the first takeLatest until it terminates, then only will run the second.

**Comment by slorber** [comment]
@yelouafi I'm not sure to understand what would be the difference between

```javascript
export default function* root() {
  yield* takeLatest(Actions.LOGIN, login);
  yield* takeLatest(Actions.LOGOUT, logout);
}
```

and

```javascript
export default function* root() {
  yield takeLatest(Actions.LOGIN, login);
  yield takeLatest(Actions.LOGOUT, logout);
}
```

**Comment by yelouafi** [comment]
@slorber from the order of execution POV there is no diff. But when testing `root` it makes a difference

For example, if we have

```javascript
function* child() {
  yield 1
  yield 2
}
```

Using `yield* child()` will delagate all next calls to `child`
```javascript
export default function* root() {
  yield* child()
  yield* child()
}

// testing
iter = root()
iter.next().value // => 1
iter.next().value // => 2
iter.next().value // => 1
iter.next().value // => 2
```

Using `yield` only evaluate and returns the expression

```javascript
export default function* root() {
  yield child()
  yield child()
}

// testing
iter = root()
iter.next().value // => result of child() => an iterator
iter.next().value // => result of child() => an iterator
```

when you yield an iterator directly like in `yield takeLatest(...)`, the middleware autoamtically resolves the yielded iterator (like it does with Promises): it runs it and waits for it to terminate and possibly return a value. But testing raw iterators could be difficult (like testing raw promises)

When you use the delegation form `yield* takeLatest()` all the yields in the child are 'copied' into the parent Saga;

**Comment by slorber** [comment]
ok so it's like flatMap for generators I guess haha

**Comment by yelouafi** [comment]
yes kind of a hardcore/brutal flatmap

**Comment by notgiorgi** [comment]
I was following the tutorial [here](https://redux-saga.github.io/redux-saga/docs/basics/UsingSagaHelpers.html) and I'm curious, whats the difference between:

```javascript
export default function* root() {
  yield helloSaga()
  yield watchIncrementAsync()
}
```
and

```javascript
export default function* root() {
  yield [
    helloSaga()
    watchIncrementAsync()
  ]
}
```

**Comment by Andarist** [comment]
The first one will is sequential - `watchIncrementAsync` will run only after `helloSaga` is done.

By `yield`ing an array to the `redux-saga` you order it to run effects (in this case those 2 iterators) in parallel. Next line after this `yield` though would be run only after both of those 2 iterators finish.

**Comment by notgiorgi** [comment]
@Andarist 
> The first one will is sequential - watchIncrementAsync will run only after helloSaga is done.
What if there was something like this

```javascript
export default function* rootSaga() {
  yield takeEvery('foo', fooHandler)
  yield takeEvery('foo', anotherFooHandler)
}
```

Will the summary of things happened after `'foo'` action is dispatched be:

- rootSaga is run
- It will run first handler
- Only after first handler which is a generator function is completed  (with all promise yields resolved delays, etc.) the second handler generator will be run

If I yielded an array of those two handlers, they would start in parallel

Correct me if I'm wrong and sorry if it's an obvious question.

**Comment by Andarist** [comment]
Not quite, you've asked before about the other thing, mentioning some random generators (`helloSaga` and `watchIncrementAsync`) and they would behave like I've described once `root` is ran.

Here you have mentioned different scenario which utilizes `takeEvery` helpers. You must understand that `redux-saga` is kinda the **interpreter** of what's being yielded, so it can interpret them.

So that being said `takeEvery` is a special thing which `redux-saga` **knows** how to interpret. It registers a 'background' task which listens for the `pattern` (in your example it's `'foo'`) and `fork`s a `worker` when it happens (in your example respectively `fooHandler` and `anotherFooHandler`).

So to sum it up when you call `sagaMiddleware.run(root)` it will register 2 listeners/tasks for `foo` action type.

And then when you dispatch `foo` it will run `fooHandler` and `anotherFooHandler`, independently of each other (those tasks do not know about each other). Probably the `fooHandler` will be resumed first but I wouldnt depend on that by any means.

**Comment by CallMeXYZ** [comment]
when dealing with multiple page sagas I'm doing sth like below.Would it be a correct and good style?Or is there a more delicate solution?
```javascript
//root
export default function* root() {
  yield [
     fork(page1),
     fork(page2),
     //and more pages
  ]
}
//page1
export default function* page1() {
    //watch actions in paralle
    yield [
        takeLatest(SOME_ACTION1, actionHandler1),
        takeLatest(SOME_ACTION2, actionHandler2), 
        //and more actions
    ]
}
```

**Comment by Andarist** [comment]
its a good structure, nothing to worry about prematurely

**Comment by panzerdp** [comment]
First of all, thanks for this awesome library!

I faced almost the same issue, but with the difference that I need to take latest action from a collection of different types.
Unfortunately `takeLatest('ACTION', handler)` works with a single type of action.

I implemented a custom effect named `takeMultipleLatest()` that accepts an object with keys as actions and values as corresponding handlers:

```javascript
function* takeMultipleLatestSaga(sagasByPatterns, ...args) {
  const patterns = Object.keys(sagasByPatterns);
  const task = yield fork(function* () {
    let lastTask;
    while (true) {
      const action = yield take(patterns);
      if (lastTask) {
        yield cancel(lastTask);
      }
      lastTask = yield fork(sagasByPatterns[action.type], ...args.concat(action));
    }
  });
  return task;
}

export function takeMultipleLatest(sagasByPatterns) {
  return call(takeMultipleLatestSaga, sagasByPatterns);
}
```
Then it can be used almost the same as `takeLatest()`, but with the difference that it supports multiple actions from which only the latest will be taken and other ones cancelled:

```javascript
  yield takeMultipleLatest({
    'ACTION1': handlerForAction1,
    'ACTION2': handlerForAction2
  });
```
I hope someone will find it useful!

**Comment by Andarist** [comment]
Could have also be written like this
```js
yield all(Object.keys(sagasByPatterns).map(pattern => takeLatest(pattern, sagasByPatterns[pattern])))
```
or by extracting it to a helper
```js
const takeMultipleLatestSaga = (sagasByPatterns, ...args) => all(Object.keys(sagasByPatterns)
  .map(pattern => takeLatest(pattern, sagasByPatterns[pattern], ...args)))
```

**Comment by panzerdp** [comment]
Hi @Andarist,

I tried you first suggestion, and unfortunately it does not cancel the tasks of different types. Maybe I'm missing something.

Thanks.

**Comment by Andarist** [comment]
ah, yeah, sorry i thought that ur use case was a little bit different, sorry for the confusion, was reading in a hurry

if you want different patterns causing different sagas, but each action still canceling any (latest) task, then ur helper is just fine :)

`takeLatest` is using regular `take` under the hood, so every pattern is allowed, if somebody needs to start the same saga for multiple action types (+ cancellation logic ofc) - `yield takeLatest([ACTION_1, ACTION_2], saga)`
